internal/controller/events.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2024 Telespazio UK.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+
+	corev1alpha1 "github.com/UKEODHP/workspace-controller/api/v1alpha1"
+	"github.com/apache/pulsar-client-go/pulsar"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type EventsReconciler struct {
+	Client
+}
+
+
+func SendWorkspaceUpdateMessage(ctx context.Context,
+	spec *corev1alpha1.WorkspaceSpec,
+	status *corev1alpha1.WorkspaceStatus,
+	pulsarURL string,) error {
+
+	log := log.FromContext(ctx)
+
+	// Send request to create these files via Pulsar
+	log.Info("Creating pulsar!")
+	// Create a Pulsar client
+	client, err := pulsar.NewClient(pulsar.ClientOptions{
```

Comment:
there will be an avoidable overhead with creating the client for each message. The singleton pattern would be good here. I can implement it easy enough, so will make an update to the PR and can close on your behalf once I have done so.